File: ruby/leet_code/402.rb
# @param {String} num
# @param {Integer} k
# @return {String}
def remove_kdigits(num, k)
  # If the length of the number is equal to k, return "0"
  return '0' if num.length == k

  # Create a stack and iterate over each character in the number
  stack = num.each_char.with_object([]) do |digit, stack|
    # While the stack is not empty, k is not zero, and the current digit is less than the last digit in the stack
    # decrement k and pop the last digit from the stack
    k, = k.pred, stack.pop until stack.empty? || k.zero? || digit.to_i >= stack.last.to_i

    # Push the current digit onto the stack
    stack << digit
  end

  # Pop k digits from the stack
  k.times { stack.pop }

  # Remove leading zeros from the stack
  stack.shift until stack.first != '0'

  # If the stack is empty, return "0", otherwise join the stack into a string and return it
  stack.empty? ? '0' : stack.join
end

File: rust/leet_code/src/402.rs
impl Solution {
    /// Removes `k` digits from `num` to create the smallest possible number.
    ///
    /// # Arguments
    ///
    /// * `num` - A string representation of the number from which digits are to be removed.
    /// * `k` - The number of digits to remove from `num`.
    ///
    /// # Returns
    ///
    /// * A string representation of the smallest possible number after removing `k` digits.
    pub fn remove_kdigits(num: String, k: i32) -> String {
        // Convert k to usize for compatibility with Vec operations
        let mut k = k as usize;

        // Initialize an empty stack to hold the digits of the number
        let mut stack: Vec<char> = Vec::new();

        // Iterate over each character in the number
        for c in num.chars() {
            // While the stack is not empty, k is greater than 0, and the current character is less than the last character in the stack
            while !stack.is_empty() && k > 0 && c < *stack.last().unwrap() {
                // Remove the last character from the stack and decrement k
                stack.pop();
                k -= 1;
            }

            // Push the current character onto the stack
            stack.push(c);
        }

        // Remove the remaining digits from the end of the number
        for _ in 0..k {
            stack.pop();
        }

        // Convert the stack into a string to construct the smallest possible number
        let mut result = stack.into_iter().collect::<String>();

        // Remove leading zeros from the number
        result = result.trim_start_matches('0').to_string();

        // If the result is an empty string, return "0"
        if result.is_empty() {
            return "0".to_string();
        }

        // Return the smallest possible number as a string
        result
    }
}
